### DISCUSSION TITLE
How to pass a subset of state to a substore

### DISCUSSION THREAD

**Discussion by ivangodfather** [pointfreeco/swift-composable-architecture#1125] [Q&A] [answered]
Lets make a quick example with "pseudo" code

```
struct User {
  let username: String
}
struct AppState {
   var user: User?
   var updateUser = UpdateUserState()
}
enum AppAction {
   case updateUser(UpdateUserAction)
}
struct UpdateUserState {
  var username = ""  // this would be binded to a TextField
}
enum UpdateUserAction {
    case performRequestToUpdateUser(username)  // this will call environment.updateUserEffect(username).map(UpdateUserAction.updatedUser)
     case updatedUser(User)
}
```

When I present the UpdateUserView I do something in these lines...

```
UpdateUserView(
	store: store.scope(
		state: \.updateUser,
		action: AppAction.updateUser
	)
)
```
I certainly get the updated user from the "appReducer" (not declared here) by simply:
```
case .updateUser(.updatedUser(let newUser)):
  state.user = newUser
```
But how I'm supposed to pass the user to the "UpdateUserView" for example to prefill the textField with the existing one from `AppState`?

**Comment by tgrapperon** [comment, marked as answer]
Hello @ivangodfather!
As often, there are several ways to achieve this.
The most common is to rely on a computed property that is updated on "get" and reinjected on "set". So, with your current model:
```swift
extension AppState {
  var updateUser: UpdateUserState {
    get {
      UpdateUserState(username: user.username)
    }
    set {
      self.user.username = newValue.username
    }
  }
}
```
This is more foolproof than having to account for each possible `username` modification.
Using this setup, a whole new `UpdateUserState` is delivered each time. It furthermore frees one stored field from your `AppState` which is always a good thing (faster automatic equatability, smaller footprint on the stack, etc.)
But this approach needs to be slightly adapted when your child state has internal or unrelated fields. For example, if your `UpdateUserState` is 
```swift
struct UpdateUserState {
  var username = ""
  var cursorPosition: CGPoint?
}
```
you will lose the `cursorPosition` value each time the state is extracted from `AppState`. The usual workaround is to use a private storage for `UpdateUserState`, and update this storage when the property is accessed/reinjected:
```swift
struct AppState {
  private var _updateUser: UpdateUserState = .init()

  var updateUser: UpdateUserState {
    get {
      var value = _updateUser
      value.username = user.username
      return value
    }
    set {
      _updateUser = newValue
      self.user.username = newValue.username
    }
  }
  …
}
```
This method is general, and can apply to almost all cases, but it could be not performant enough when your child state is a large collection. If performance is problematic, you can then either:
- watch for actions that are susceptible to change the `User` in `AppState`, and update your child state accordingly, as you suggested.
- watch the `user` value directly if it is `Equatable`, and update the child state only when it changes. This can be conveniently achieved using some higher-order reducer similar to [`onChange` from isowords](https://github.com/pointfreeco/isowords/blob/main/Sources/TcaHelpers/OnChange.swift). In this case, you only have to tap `.onChange` at the end of your `appReducer`:
```swift
let appReducer = Reducer<…> {
  …
}
.onChange(\user.username) { newValue, state, action, environment in
  state.updateUser.username = newValue
}
```
Each method has its pros and cons, and you may have to switch from one to another as your app evolves and you hit their own limitations. There is no silver bullet to my knowledge.

**Reply by ivangodfather** [reply]
Hello @tgrapperon ! Thanks (again) for your help!. 
Maybe I wasn't clear enough with the question, but I sort it out in a different manner.

 In short: I still have my `UpdateUserState` as a normal property property on `AppState`, but also I nil that out. Then when I'm going to present the view (via .sheet) I set first the value for this `state.updateUser = UpdateUserState(user: state.user); state.isUpdateUserPresented = true`. At this point the user is in the Child, so I can read values. So I created a new action on the child `onAppear`, where I update the child username for whatever the username is. Then after calling the environment.UpdateUserEffect I can read that new user in the AppState via `case .updateUser(. updatedUser(.success(let user)))` and then update my `state.user = user`. I know when I asked I talked about "username" and not the whole object, but should be the same approax. Can you give me your thougths on this?

 I'll try to expand here as much as I can but from my little understanding seems reasonable (at least seems working)

```
struct AppState {
  var updateUser: UpdateUserState?
  var isUpdateUserViewPresented = false
  var user: User?
}
enum AppAction {
    case setUpdateUserSheet(isPresented: Bool)
   case .updateUser(UpdateUserAction)`
}

let appReducer = Reducer<...> { state, action, env in
		case .setUpdateUserSheet(let isPresented):
			state.updateUser = .init(user: state.user)
			state.isUpdateUserViewPresented = isPresented
			return .none
		case .updateUser(.userUpdated(.success(let user))):
			state.user = user
			state.isUpdateUserViewPresented = false
			return .none
}
```

Then in my AppView...
```
			.sheet(
				isPresented: viewStore.binding(
					get: \.isUpdateUserViewPresented,
					send: AppAction.setUpdateUserSheet(isPresented:)
				)
			) {
				IfLetStore(store.scope(state: \.updateUser, action: AppAction.updateUser)) { store in
					UpdateUserView(store: store)
				}

			}
```

And finally on my childView I call onAppear where I set in the reducer the username from the user.username
```
	case .onAppear:
		state.username = state.user?.userName ?? ""
		return .none
```

**Reply by tgrapperon** [reply]
OK, a few comments: I think your `isUpdateUserViewPresented` property should be a read-only computed property that returns `updateUser == nil`. Thus, you don't have to manage information twice. When `updateUser` is not nil, the modal sheet is presented. Your current reducer will set some non-nil `updateUser` and `isUpdateUserViewPresented == false` if it receives `.setUpdateUserSheet(isPresented:false)`. This state is ambiguous, but you can structure your model to prevent this from being possible. If you receive `.setUpdateUserSheet(isPresented:false)`, you set `updateUser` to nil, if you receive `isPresented:true`, you assign a user with the current `user.userName` value. Thus, when the sheet appears, the `updateUser` state has the correct value. 

But this value won't be synchronized with `.user.userName`: the changes from the `TextField` will stay internal to `updateUser`. One solution is to catch binding actions from the app reducer and assign `updateUser.name` to `user.userName`. There is one issue with this however: it forces you to acknowledge the internals of your `UpdateUser` feature. If you add a "reset" button that clears the field, you will need to observe this action from the app state too in order to propagate this change to `user`. Same thing if the changes come from a delayed effect from your environment.

I would advise to set up your model so you only have a single source of truth in your state. If the exact same information appears twice, you should try to implement automatic synchronization between both values, for example using the computed property trick, which is the strongest guarantee of all as it doesn't depend on a reducer to be effective. Reducer-enforced sync offers a weaker guarantee, but can also be more efficient in terms of speed. If your model is automatically synchronized, no matter what changes the value, it is always consistent with what the user (or your app) is seeing.

About `.onAppear`: if your child reducer does nothing itself from this action, you're creating a coupling between your child and the outer world. This is suboptimal: the child relies on some external entity it doesn't know to realize some work. Ideally, the child should be able to work in isolation. However, if the child retrieves itself the user info from the environment in its `.onAppear`, this information will automatically flow up in the app state through "automatic" synchronization. If you don't want this to happen and provide the child with already up-to-date information, you can use another child reducer that does nothing for `.onAppear` in your pullback. To this mean, you can use something similar to the [filter reducer from isowords](https://github.com/pointfreeco/isowords/blob/main/Sources/TcaHelpers/FilterReducer.swift) to derive a child reducer that does nothing on `.onAppear` from one that does.

I'm not sure I'm understanding correctly your issue otherwise. In this case, are you able to publish a basic project exhibiting the issue you have?

**Reply by ivangodfather** [reply]
@tgrapperon 
https://github.com/ivangodfather/TestTCA

**Reply by tgrapperon** [reply]
Thanks @ivangodfather! I've tested the code and it seems to work as expected on my side. I've refactored a few things while trying to preserve most of it and created a PR. Can you point where the issue is on your side?

**Reply by ivangodfather** [reply]
@tgrapperon thank you! Let me go over the PR. There is no issue on my side, just wanted to show how I "solved" my own issue and wanted to double check with you that I was not doing anything super wrong.

**Reply by ivangodfather** [reply]
Everything makes a lot of sense besides adding the "username" into the initializer. If so, I would rather call it "placeholder" so the caller is aware what that actually is (he can put whatever wants there), whereas using my approach I set privately the placeholder always to the username.

**Reply by tgrapperon** [reply]
The placeholder can also be something you only display in the UI when `userName` is empty.
But I'm glad it helped. In this specific case, I would personally use non-optional `User`s, and a computed property with private storage for `updateUser` state:
```swift
var user: User
var _updateUser: UpdateUserState?

var updateUser: UpdateUserState? {
  get {
    guard var value = _updateUser else { return nil }
    value.user = user
    return value
  }
  set {
    _updateUser = newValue
    if let newValue = newValue {
      user = newValue.user
    }
  }
}
```
In this way, there is only one source of truth: the app's `User`, and you don't have to remember to update its value in the app's reducer when you receive some `.updateUser` action. You need private storage because the `isLoading` value need to be stored somewhere, but your `UpdateUser` feature always receives an up to date value for `User`.

**Reply by ivangodfather** [reply]
@tgrapperon I have a question regarding this last code (I saw also in your previous example), wouldnt the "set" block be called each time you modify any property of the state inside `UpdateUserState` ? like when you are setting `state.isLoading = false`. So you'll be updating the user even if you are only interacting with the view (like updating the textfield, if this view has toggles, etc...).

**Reply by tgrapperon** [reply]
Yes, it is, and that's why it's important to carve-out the piece of state you want to observe in your `ViewStore`s. I was very worried about this for a long time, but experience showed that it is not really an issue most of the time. And in cases it would matter, you can always fallback to the reducer being in charge of the state synchronization.
